src/command.rs

```diff
@@ -0,0 +1,105 @@
+// Copyright 2020 Contributors to the Parsec project.
+// SPDX-License-Identifier: Apache-2.0
+
+//! Command implementations. Interacts with parsec-client-rust.
+
+use crate::cli::{Action, CommandLineContext};
+use crate::error::ParsecCliError;
+use anyhow::{Context, Result};
+use parsec_client::auth::AuthenticationData;
+use parsec_client::core::interface::requests::ProviderID;
+use parsec_client::BasicClient;
+
+fn get_unauthenticated_client() -> BasicClient {
+    BasicClient::new(AuthenticationData::None)
+}
+
+/// Gets the list of availble opcodes for a given provider and outputs them.
+fn list_opcodes(
+    provider: &str,
+    _command_line_context: &CommandLineContext,
+) -> Result<(), ParsecCliError> {
+    let client = get_unauthenticated_client();
+
+    let desired_provider = match provider {
+        "core" => ProviderID::Core,
+        "mbedcrypto" => ProviderID::MbedCrypto,
+        "pkcs11" => ProviderID::Pkcs11,
+        "tpm" => ProviderID::Tpm,
+        _ => unreachable!(),
+    };
+
+    let provider_opcodes = client
+        .list_opcodes(desired_provider)
+        .context("Failed to list opcodes.")?;
+
+    e_info!("Available opcodes for provider {:?}:", desired_provider);
+    for provider_opcode in provider_opcodes {
+        print_colored!(Blue, "*");
+        println!(" {:?}", provider_opcode);
+    }
+    Ok(())
+}
+
+/// Gets the list of available providers and outputs them.
+fn list_providers(_command_line_context: &CommandLineContext) -> Result<(), ParsecCliError> {
+    let client = get_unauthenticated_client();
+    let available_providers = client
+        .list_providers()
+        .context("Failed to list providers.")?;
+
+    e_info!("Available providers:");
+    for provider in available_providers {
+        print!("[");
+        print_colored!(Cyan, "0x{:02x} ({})", provider.id as u32, provider.id);
+        println!("]");
+
+        print_colored!(Yellow, "Description");
```

Comment:
If you keep this system for colouring output, you should probably write down somewhere what each colour represents (for example, why the details here are yellow - I would've assumed yellow is more of a warning colour).

Reply:
I'm not saying yellow is wrong, btw, just that it should be clear what the colours are used for

Reply:
That's a good idea -- so far, the format is:

* Yellow to represent field names -- i.e. 'description', 'uuid'.
* Cyan for titles.
* Blue for information.
* Red for errors (and warnings?).
* Green for success messages.

Reply:
If we agree on a list like this, would it make sense to make separate macros for each of these to handle both the colour and the printing inside the macro? So for example, instead of
```rust
print_colored!(Yellow, "Description");
println!(": {}", provider.description);
```
you'd get the same result via
```rust
print_field!("Description", "{}", provider.description);
```
Apologies if this is a bit over the top 😅